### ISSUE TITLE
How number of classes in training impact on FPS

### ISSUE THREAD

**Issue by danimod92** [AlexeyAB/darknet#4020]
Hello,
I am new about Computer Vision and Yolo world. I did train a network on darknet by using Yolov3 tiny (since I can't run Yolov3 because of my poor GPU). I took a subset from COCO dataset because I wanted to try the network in order to detect only the class 'Persons'. I followed the guide for configuring properly the configure and I used Batch = 64 and subdivision = 16
At the end of training, the average loss stuck around 2.9 (after 2000 iteratation, since I set max batch = 2000*classes), indeed the accuracy is not that high.
I tested it on some images and videos, getting around 50-60FPS and on Webcam about 16FPS.
But on my project purpose I have to test it on a stream video cam. This way the FPS drops till 3-4FPS.but a very poor accuracy.
My questions are: Is the number of classes I use to train the network impact on the FPS or the result would have been the same if I e.g. trained the net with 30-50 classes?
And How could I improve my FPS? I guess one solution would be of course getting a larget GPU

**Comment by AlexeyAB** [comment]
> My questions are: Is the number of classes I use to train the network impact on the FPS or the result would have been the same if I e.g. trained the net with 30-50 classes?

No.

> And How could I improve my FPS?

Train this cfg-file: https://github.com/AlexeyAB/darknet/files/3580764/yolo_v3_tiny_pan3_aa_ae_mixup_scale_giou.cfg.txt

**Comment by danimod92** [comment]
Thank you for the answer @AlexeyAB. As far as I understood the number of classes have an impact on the training processing time, becaue of the larger dataset and the tensor size (since It depends on #classes). Why does not influence the FPS speed?

Do I have to retrain the network with this cfg file?

**Comment by AlexeyAB** [comment]
> tensor size (since It depends on #classes). Why does not influence the FPS speed?

Only size of one conv-layer before [yolo] layer depends on classes. So there is < 1% of impact.

Change cfg-file

**Comment by danimod92** [comment]
I tried to use such cfg file but at the end of the training (with an avg loss of  1.78) now it does not detect anything, no bounding boxes (only when I change the threshold to 0.1 e.g.). It just detects with an accuracy of 1%. Which is weird because I did use the same exact dataset as the previous one.
This is my obj.data which I placed in darknet/data

classes= 1
train  = data/train.txt
valid  = data/test.txt
names = data/obj.names
backup = backup/

- Obj.names contains only one line with the label 'person' I want to detect
- The images and their labels in .txt format are in data/obj folder
- My cfg file is (I changed batch and subdivision because I got Out of Memory error. Filters are the sames as you sent me, because of the formula (#classes + 1)* number_masks
[net]
# Testing
#batch=1
#subdivisions=1
# Training
batch=64
subdivisions=64
width=544
height=544
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1
mixup=1
#blur=5


learning_rate=0.001
burn_in=1000
max_batches = 2000

policy=sgdr
sgdr_cycle=1000
sgdr_mult=2
steps=1600, 1800
#scales=1, 1, 0.1, 0.1


[convolutional]
batch_normalize=1
filters=16
size=3
stride=1
pad=1
activation=swish

#[convolutional]
#batch_normalize=1
#filters=16
#size=3
#stride=2
#pad=1
#activation=swish

# remove
[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=swish

[maxpool]
size=2
stride=2
antialiasing=1

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=swish

[maxpool]
size=2
stride=2
antialiasing=1

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=swish

[maxpool]
size=2
stride=2
antialiasing=1

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=swish

[maxpool]
size=2
stride=2
antialiasing=1

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=swish

[maxpool]
size=2
stride=1

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=swish

###########

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=swish

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=swish
assisted_excitation=4000


########### to [yolo-3]


[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=swish

[upsample]
stride=2

[route]
layers = -1, 8

###########

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=swish


[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=swish

########### to [yolo-2]


[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=swish

[upsample]
stride=2

[route]
layers = -1, 6

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=swish

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=swish


########### features of different layers


[route]
layers=2

[maxpool]
size=16
stride=16

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=swish

[route]
layers=4

[maxpool]
size=8
stride=8

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=swish

[route]
layers=4

[maxpool]
size=8
stride=4
stride_x=4
stride_y=8

[convolutional]
batch_normalize=1
filters=64
size=1
stride=2
stride_x=2
stride_y=1
pad=1
activation=swish

[route]
layers=4

[maxpool]
size=8
stride=8
stride_x=8
stride_y=4

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
stride_x=1
stride_y=2
pad=1
activation=swish

[route]
layers=4

[maxpool]
size=8
stride=8

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=swish

[route]
layers=6

[maxpool]
size=4
stride=4

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=swish

[route]
layers=6

[maxpool]
size=4
stride=2
stride_x=2
stride_y=4

[convolutional]
batch_normalize=1
filters=64
size=1
stride=2
stride_x=2
stride_y=1
pad=1
activation=swish

[route]
layers=6

[maxpool]
size=4
stride=4
stride_x=4
stride_y=2

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
stride_x=1
stride_y=2
pad=1
activation=swish

[route]
layers=8

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=swish

[route]
layers=10

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=swish

[route]
layers=-1, -3, -6, -9, -12, -15, -18, -21, -24, -27

[maxpool]
maxpool_depth=1
out_channels=64
stride=1
size=1


########### [yolo-1]


[upsample]
stride=4

[route]
layers = -1,24

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=swish


[convolutional]
size=1
stride=1
pad=1
filters=30
activation=linear

[yolo]
mask = 0,1,2,3,4
anchors = 8,8, 10,13, 16,30, 33,23,  32,32, 30,61, 62,45, 64,64,  59,119, 116,90, 156,198, 373,326
classes=1
num=12
jitter=.3
ignore_thresh = .7
truth_thresh = 1
iou_normalizer=0.25
cls_normalizer=1.0
iou_loss=giou
scale_x_y = 1.05
random=0




########### [yolo-2]

[route]
layers = -6

[upsample]
stride=2

[route]
layers = -1,19

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=swish

[convolutional]
size=1
stride=1
pad=1
filters=30
activation=linear

[yolo]
mask = 4,5,6,7,8
anchors = 8,8, 10,13, 16,30, 33,23,  32,32, 30,61, 62,45, 64,64,  59,119, 116,90, 156,198, 373,326
classes=1
num=12
jitter=.3
ignore_thresh = .7
truth_thresh = 1
iou_normalizer=0.25
cls_normalizer=1.0
iou_loss=giou
scale_x_y = 1.1
random=0




########### [yolo-3]

[route]
layers = -12


[route]
layers = -1,14

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=swish

[convolutional]
size=1
stride=1
pad=1
filters=24
activation=linear



[yolo]
mask = 8,9,10,11
anchors = 8,8, 10,13, 16,30, 33,23,  32,32, 30,61, 62,45, 59,119,   80,80, 116,90, 156,198, 373,326
classes=1
num=12
jitter=.3
ignore_thresh = .7
truth_thresh = 1
iou_normalizer=0.25
cls_normalizer=1.0
iou_loss=giou
scale_x_y = 1.2
random=0

**Comment by AlexeyAB** [comment]
1. How many iterations did you train?
2. Show screenshot of mAP calculation

**Comment by danimod92** [comment]
I trained for 2000 iterations. Previously I trained for 3000 iterations and the result was the same, always getting same average loss, no less than 1.5
![chart](https://user-images.githubusercontent.com/45860693/66254824-08234180-e77c-11e9-997a-70bd7756e749.png)

I run the code
`./darknet detector map data/obj.data yolo_v3_gpu.cfg backup/yolo_v3_gpu_final.weights 
`
for getting the map and that's what I got:
![MAP](https://user-images.githubusercontent.com/45860693/66254982-36a21c00-e77e-11e9-81f6-003203621e79.png)

**Comment by AlexeyAB** [comment]
> I have another doubt: should I fill the test.txt file as the train.txt with the relative path of the images I will use as a test?

I don't understand.

Train at least 4000 iterations.

Train with flag `-map` at the end of training command, and train while the mAP increases (red line on the chart).
